pmd-core/src/main/java/net/sourceforge/pmd/ruledef/RuleFactory.java

```diff
@@ -0,0 +1,412 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.ruledef;
+
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.CLASS;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.DEPRECATED;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.DFA;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.EXTERNAL_INFO_URL;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.LANGUAGE;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.MAXIMUM_LANGUAGE_VERSION;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.MESSAGE;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.METRICS;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.MINIMUM_LANGUAGE_VERSION;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.NAME;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.PROPERTY;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.SINCE;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.TYPERESOLUTION;
+import static net.sourceforge.pmd.ruledef.RulesetSchemaConstants.VALUE;
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.apache.commons.lang3.EnumUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+import net.sourceforge.pmd.Rule;
+import net.sourceforge.pmd.RulePriority;
+import net.sourceforge.pmd.lang.rule.RuleReference;
+import net.sourceforge.pmd.properties.AbstractPropertyDescriptorFactory;
+import net.sourceforge.pmd.properties.PropertyDescriptor;
+import net.sourceforge.pmd.properties.PropertyDescriptorFactory;
+import net.sourceforge.pmd.properties.PropertyDescriptorField;
+import net.sourceforge.pmd.properties.PropertyDescriptorUtil;
+
+/**
+ * Builds rules from rule XML nodes.
+ *
+ * @author Clément Fournier
+ * @since 6.0.0
+ */
+public class RuleFactory {
+
+    public static final RuleFactory INSTANCE = new RuleFactory();
```

Comment:
You're definitely right. I thought that it could work that way since it had no instance variables, but if we one day add instance variables then we'd have to break down the API to make the change.